Sources/ShapedArray/Basic.swift

```diff
@@ -49,6 +53,115 @@ extension ShapedArray {
         self.reshaped(to: -1)
     }
 
+    // helper data structure for subarray generation
+    @usableFromInline
+    internal struct SubarrayIndices {
+        @usableFromInline
+        let start: [Int]
+        @usableFromInline
+        let end: [Int]
+    }
+
+    // helper function for subarray indices generation
+    /// takes a shape for the original array, a shape for the subarray, 
+    /// and an axis on which the original array would be divided 
+    /// and returns an array of start and end indices for each subarray
+
+    @usableFromInline
+    internal func _calculateSubarrayIndices(shape: [Int], subarrayShape: [Int], axis: Int) -> [SubarrayIndices] {
+        let numberOfSubarrays = shape[axis] / subarrayShape[axis]
+        
+        var subarrays = [SubarrayIndices]()
+        
+        for i in 0..<numberOfSubarrays {
+            var startIndices = [Int](repeating: 0, count: shape.count)
+            var endIndices = shape
+            
+            startIndices[axis] = i * subarrayShape[axis]
+            endIndices[axis] = startIndices[axis] + subarrayShape[axis]
+            
+            subarrays.append(SubarrayIndices(start: startIndices, end: endIndices))
+        }
```

Comment:
this may look a bit nicer with a map
```swift
let subarrays = (0..<numberSubarrays).map { i in
    var startIndices = [Int](repeating: 0, count: shape.count)
    var endIndices = shape

    startIndices[axis] = i * subarrayShape[axis]
    endIndices[axis] = startIndices[axis] + subarrayShape[axis]

    return SubarrayIndices(start: startIndices, end: endIndices)
}
```